pageserver/src/page_image_cache.rs

```diff
@@ -0,0 +1,307 @@
+                // cache hit
+                match &cache.pages[index].state {
+                    PageImageState::Loaded(cached_page) => {
+                        // Move to the head of LRU list
+                        let page = cached_page.clone();
+                        cache.unlink(index);
+                        cache.link_after(0, index);
+                        return Ok(page);
+                    }
+                    PageImageState::Loading(event) => {
+                        // Create event on which to sleep if not yet assigned
+                        let cv = match event {
+                            None => {
+                                let cv = Arc::new(Condvar::new());
+                                cache.pages[index].state =
+                                    PageImageState::Loading(Some(cv.clone()));
+                                cv
+                            }
+                            Some(cv) => cv.clone(),
+                        };
+                        cache = cv.wait(cache).unwrap();
+                        // Retry lookup
+                        continue 'lookup;
+                    }
+                    PageImageState::Vacant => bail!("Vacant entry is not expected here"),
+                };
+            }
+            index = cache.pages[index].collision;
+        }
+        // Cache miss
+        index = cache.free_list;
+        if index == 0 {
+            // no free items
+            let victim = cache.pages[0].prev; // take least recently used element from the tail of LRU list
+            assert!(victim != 0);
+            // Remove victim from hash table
+            let h = hash(&cache.pages[victim].key) % cache.hash_table.len();
+            index = cache.hash_table[h];
+            let mut prev = 0usize;
+            while index != victim {
+                assert!(index != 0);
+                prev = index;
+                index = cache.pages[index].collision;
+            }
+            if prev == 0 {
+                cache.hash_table[h] = cache.pages[victim].collision;
+            } else {
+                cache.pages[prev].collision = cache.pages[victim].collision;
+            }
+            // and from LRU list
+            cache.unlink(victim);
+
+            index = victim;
+        } else {
+            // Use next free item
+            cache.free_list = cache.pages[index].next;
+        }
+        // Make is_empty(index) == true. If entry is removed in process of loaded,
+        // it will be updated so that !is_empty(index)
+        cache.prune(index);
+
+        // Insert in hash table
+        cache.pages[index].collision = cache.hash_table[h];
+        cache.hash_table[h] = index;
+
+        cache.pages[index].key = key;
+        cache.pages[index].state = PageImageState::Loading(None);
+        drop(cache); //release lock
+
+        // Load page
+        let page = Arc::new(timeline.get_rel_page_at_lsn(rel, blkno, lsn, true)?);
```

Comment:
If the `timeline.get_rel_page_at_lsn` returns an error, there might be a `Arc<CondVar>` which has been created which will now have some thread waiting on it. If the `?` is used here, then the waiters on the CondVar are never woken up, for example to re-try the read.

Reply:
Well, this is true. I first try to implement passing error to waiting thread. 
So instead of `Loaded(Arc<Bytes>)` have ` Loaded(Result<Arc<Bytes>>)`
But anyhow::Error is non-cloneable. So I failed to fight with Rust trying to clone result.

Alternatively I can have Arc<Result<Bytes>> but then it has to be somehow transformed into Arc<Bytes>.
So I gave up. Actually there are no valid reasons for Neon to load page failures. So error here actually means some critical error and may be the best reaction on this error will be restart of pageserver.

In any case, if any body can propose some elegant solution of the problem (propagating error to all waiting threads),
I will be glad to refactor this code,

Reply:
Thank you.
I thought about using Option as error indicator but you did it for me:)